QueryEngine/RelAlgOptimizer.cpp

```diff
@@ -760,11 +762,21 @@ std::vector<std::unique_ptr<const RexAgg>> renumber_rex_aggs(std::vector<std::un
                                                              const std::unordered_map<size_t, size_t>& new_numbering) {
   std::vector<std::unique_ptr<const RexAgg>> new_exprs;
   for (auto& expr : agg_exprs) {
-    auto old_idx = expr->getOperand();
-    auto idx_it = new_numbering.find(old_idx);
-    if (idx_it != new_numbering.end()) {
-      new_exprs.push_back(
-          boost::make_unique<RexAgg>(expr->getKind(), expr->isDistinct(), expr->getType(), idx_it->second));
+    bool need_renumber = false;
```

Comment:
Couldn't we just ignore the second argument and undo most of the changes here? It's always a literal anyway.

Reply:
@asuhan I've seen you saying the same in the ticket, but wasn't sure why should I handle constant expressions differently. Perhaps, I don't completely understand how renumbering works.
Caught a cold a little bit yesterday, will look into it again as soon as I get better, probably day or two.

Reply:
Even if literals indeed never get renumbered, can't use the old code because rex size now may be between 0 and 2 (0 operands for count(*), it was '-1' operand idx before).
If you don't like unnecessary std::vector allocations and redundant check in hashmap for literal idx, here how I can rewrite the function:

```c
std::vector<std::unique_ptr<const RexAgg>> renumber_rex_aggs(std::vector<std::unique_ptr<const RexAgg>>& agg_exprs,
                                                             const std::unordered_map<size_t, size_t>& new_numbering) {
  std::vector<std::unique_ptr<const RexAgg>> new_exprs;
  for (auto& expr : agg_exprs) {
    if (expr->size() >= 1) {
      auto old_idx = expr->getOperand(0);
      auto idx_it = new_numbering.find(old_idx);
      if (idx_it != new_numbering.end()) {
        std::vector<size_t> operands;
        operands.push_back(idx_it->second);
        if (expr->size() == 2) {
          operands.push_back(expr->getOperand(1));
        }
        new_exprs.push_back(boost::make_unique<RexAgg>(expr->getKind(), expr->isDistinct(), expr->getType(), operands));
        continue;
      }
    }
    new_exprs.push_back(std::move(expr));
  }
  return new_exprs;
}
```

I'd probably prefer the current version because:
1. it will work whatever happens in other parts of code (if literals get renumbered some other day, or perhaps we add another aggregate with two arguments)
2. this function is far from hot path, so one allocation and one lookup probably won't make too big a difference.

@asuhan what do you think?

Reply:
I'm not convinced that calling renumbering on literals is a legitimate use of this code. I'd rather have it fail quickly if the caller attempts something dumb. I have no other concerns, of course the allocation and lookup you do here are completely fine. Anyway, the patch looks good overall and I'm not going to hold it because this renumbering change, at least it looks correct.